core/chapters/c12_dictionaries.py

```diff
@@ -0,0 +1,510 @@
+            assert_equal(substitute(original), expected)
+
+
+    ##TODO: Is this supposed to be an exercise or a verbatim step?
+    class dna_part2(VerbatimStep):
+        """
+Now we can use dictionaries to make this code both shorter and more general so it can be used for other purposes.
+
+Your job is to add another argument to the `substitute` function: a dictionary called `d`.
+
+The keys of `d` represent characters
+in the first argument `string` that should be replaced by the corresponding values of `d`. For example, `'A': 'T'`
+means that `A` should be replaced by `T`:
+
+    __copyable__
+    __program_indented__
+        """
+        
+        def program(self):
+            def substitute(string, d):
+                ...
+
+            original = 'AGTAGCGTCCTTAGTTACAGGATGGCTTAT'
+            expected = 'TCATCGCAGGAATCAATGTCCTACCGAATA'
+            assert_equal(substitute(original, {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}), expected)
+
+
+    class cryptography(VerbatimStep):
+        """
+Nice! Here's an example of how this function can also be used to encrypt and decrypt secret messages:
+
+    __copyable__
+    __program_indented__
+        """
+
+        def program(self):
+            plaintext = 'helloworld'
+            encrypted = 'qpeefifmez'
+            letters = {'h': 'q', 'e': 'p', 'l': 'e', 'o': 'f', 'w': 'i', 'r': 'm', 'd': 'z'}
+            reverse = {'q': 'h', 'p': 'e', 'e': 'l', 'f': 'o', 'i': 'w', 'm': 'r', 'z': 'd'}
+            assert_equal(substitute(plaintext, letters), encrypted)
+            assert_equal(substitute(encrypted, reverse), plaintext)
+
+
+    final_text = """
+The same function works in both directions, we just need to pass it different dictionaries.
+
+The two dictionaries are almost the same, we just swap around the key and value in each pair.
+So to encrypt, we replace `e` with `p`, and to decrypt we change `p` back to `e`.
+
+Note that `'e'` is both a key and a value in `letters`.
+
+Looking up `letters['e']` means that we're asking about `'e'` as a *key*, so it gives `'p'`.
+Remember, we can't use `letters` to ask which key is associated with `'e'` as a *value*.
+But in this case we can use the other dictionary for that: `reverse['e']` gives `'l'`,
+and `letters['l']` gives `'e'` again.
+
+Soon you'll write a function to create a dictionary like `reverse` automatically,
+i.e. `reverse = swap_keys_values(letters)`."""
+
+class DictionaryKeysAndValues(Page):
+    class introducing_keys(Step):
+        """
+Copy this code into the editor, then change `print(quantities)` to `print(quantities.keys())`, and run the whole program.
+
+    __copyable__
+    __no_auto_translate__
+
+    quantities = {'apple': 1, 'cat': 10}
+    print(quantities)
+
+
+        """
+
+        program = """
+        quantities = {'apple': 1, 'cat': 10}
+        print(quantities.keys())
+        """
+
+        def check(self):
+            return "dict_keys(['apple', 'cat'])" in self.result
```

Comment:
```suggestion
            return "dict_keys([" in self.result
```

This method of checking is a good idea here. This change is so that it will still work when the exercise is translated, whether that's automatic or not.